runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -12080,3 +12103,267 @@ J9::Z::TreeEvaluator::tabortEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+
+   TR::RegisterDependencyConditions* deps = generateRegisterDependencyConditions(0, 7, cg);
+   deps->addPostCondition(resultReg, TR::RealRegister::AssignAny);
+   deps->addPostCondition(locationReg, TR::RealRegister::AssignAny);
+   deps->addPostCondition(dummyRegForRA, cg->getReturnAddressRegister());
+   deps->addPostCondition(dummyRegForEntry, cg->getEntryPointRegister());
+
+   // Load address of the target field, to be used by GC evacuation helper.
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, startICFLabel);
+   startICFLabel->setStartInternalControlFlow();
+
+   int32_t shiftAmount = TR::Compiler->om.compressedReferenceShift();
+   bool shouldShift = (shiftAmount != 0);
+   TR::InstOpCode::Mnemonic loadOpCode = isCompressedRef ? TR::InstOpCode::LLGF: TR::InstOpCode::LG;
+
+   if (fieldUnresolved)
+      {
+      generateRXInstruction(cg, TR::InstOpCode::LA, node, locationReg, loadMemRef);
+      generateRXInstruction(cg, loadOpCode, node, resultReg, generateS390MemoryReference(locationReg, 0, cg));
+      }
+   else
+      {
+      generateRXInstruction(cg, loadOpCode, node, resultReg, loadMemRef);
+      }
+   if (shouldShift)
+      {
+      generateRSInstruction(cg, TR::InstOpCode::SLLG, node, resultReg, resultReg, shiftAmount);
+      }
+
+   deps->addAssignAnyPostCondOnMemRef(loadMemRef);
+
+   TR::Register* vmReg = cg->getLinkage()->getMethodMetaDataRealRegister();
+   TR::MemoryReference* baseMemRef = generateS390MemoryReference(vmReg, TR::Compiler->vm.thisThreadGetEvacuateBaseAddressOffset(comp), cg);
+   TR::MemoryReference* topMemRef  = generateS390MemoryReference(vmReg, TR::Compiler->vm.thisThreadGetEvacuateTopAddressOffset(comp), cg);
+
+   // Range check with implicit software CS status check.
+   generateRXInstruction(cg, TR::InstOpCode::CLG, node, resultReg, baseMemRef);
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BL, node, doneLabel);
+   generateRXInstruction(cg, TR::InstOpCode::CLG, node, resultReg, topMemRef);
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BH, node, doneLabel);
+   cg->generateDebugCounter(TR::DebugCounter::debugCounterName(cg->comp(), "readBar/helperCall"), 1, TR::DebugCounter::Cheap);
+   if (!fieldUnresolved)
+      {
+      generateRXInstruction(cg, TR::InstOpCode::LA, node, locationReg, generateS390MemoryReference(*loadMemRef, 0, cg));
+      }
+
+   TR::MemoryReference* gsOperandAddressMemRef = generateS390MemoryReference(vmReg, TR::Compiler->vm.thisThreadGetGSOperandAddressOffset(cg->comp()), cg);
+   TR::MemoryReference* gsHandlerAddressMemRef = generateS390MemoryReference(vmReg, TR::Compiler->vm.thisThreadGetGSHandlerAddressOffset(cg->comp()), cg);
+   generateRXInstruction(cg, TR::InstOpCode::STG, node, locationReg,  gsOperandAddressMemRef);
+   generateRXInstruction(cg, TR::InstOpCode::LG, node, raReg, gsHandlerAddressMemRef);
+   generateRRInstruction(cg, TR::InstOpCode::BASR, node, raReg, raReg);
```

Comment:
I don't think we need to load the `gsHandlerAddress` here. We dispatch the call via the `BAS` instruction.

Reply:
911a46b009 changed `LG + BASR` to `BAS`

Reply:
I switched to `S390HelperCallSnippet` which is just a BRC to helper call snippet. `BAS` is longer used. 
In ff33d584b5,  I also moved `handleSoftwareReadBarrier` to the JIT code and stopped registering it to `gsHandlerAddress`.